Issue 6574 - Move pblock access from switch to func table

mreynolds389 opened the pull request:
Description

Currently there are over 300 switch cases for getting/setting pblock parameters.

Instead we should have separate functions for each parameter, and use a callback function pointer table.

This also combines all the parameter definitions to a single ordered list. This will make future changes much easier to implement.

Relates: https://github.com/389ds/389-ds-base/issues/6574

Firstyear reviewed (approved):
Not every day you get a 5 to 8% improvement! That's amazing to see :)

droideck commented on ldap/servers/slapd/pblock.c:
```diff
@@ -368,3865 +365,5944 @@ slapi_pblock_clone(Slapi_PBlock *pb)
 {
+    Slapi_Backend *be = pblock->pb_backend;
 
-#ifdef PBLOCK_ANALYTICS
-    pblock_analytics_record(pblock, arg);
-#endif
```
This analytics code (and the one in `_set`) were removed. Should we still keep it? Or it's not used?

mreynolds389 replied:
> This analytics code (and the one in `_set`) were removed. Should we still keep it? Or it's not used?

I'm not 100% what these are for.  I assumed it was for measuring how we accessed the pblock, but I have a feeling this is no longer needed.  Perhaps we can remove it in a separate PR?

droideck commented on ldap/servers/slapd/slapi-plugin-compat4.h:
```diff
@@ -31,16 +31,6 @@ extern "C" {
-#define SLAPI_REQUESTOR_ISUPDATEDN SLAPI_IS_REPLICATED_OPERATION
-#define SLAPI_CONN_CLIENTIP 145 /* use SLAPI_CONN_CLIENTNETADDR */
-#define SLAPI_CONN_SERVERIP 146 /* use SLAPI_CONN_SERVERNETADDR */
-
```
I think we should mention the removal in the commit message.

But also, we have some leftovers in `test/libslapd/pblock/pblock_accessors*` and a bit in `_slapi_pblock_deprecated` in `ldap/servers/slapd/pblock_v3.h`.

IMO, it'll be cleaner to remove the code in a separate PR.

mreynolds389 replied:
I left this as is with the new commit

droideck reviewed (approved):
Besides a couple of minor issues, it looks like a simple but powerful fix! Nice find!

tbordaz reviewed (commented):
The change looks good and  perf improvement very nice. IIUC a former customer plugin will require to be rebuilt, correct ?

progier389 commented on ldap/servers/slapd/slapi-plugin.h:
```diff
@@ -6977,50 +6974,6 @@ slapi_timer_result slapi_timespec_expire_check(struct timespec *expire);
                                          * into tombstone
                                          */
 
-/*
```
I do not like the massive change in slapi-plugin.h 
  (looks like a risk of breaking existing custom plugin
    And after a closer look as you changed the #define values, it will definitively break the binary compatibility
     with existing custom plugin.
IMHO we should not do this. we should be able to split the switches without changing slapi-plugin.h

tbordaz replied:
an option would be to have a conversion table that convert a former define value into an index into the callback table

progier389 replied:
FYI, or you can do a simple table:
 here is a python script that generate a skeleton of what it would looks like:
[split_pblock_switches.py.txt](https://github.com/user-attachments/files/18657166/split_pblock_switches.py.txt)

mreynolds389 committed:
Issue 6574 - Move pblock access from switch to func table

Description

Currently there are over 300 switch cases for getting/setting pblock parameters.

Instead we should have separate functions for each parameter, and use a callback
function pointer table.

Relates: https://github.com/389ds/389-ds-base/issues/6574

Reviewed by: spichugi, firstyear, progier, tbordaz (Thanks!!!!)

droideck commented on ldap/servers/slapd/pblock.c:
```diff
@@ -4571,3 +6729,3949 @@ bind_credentials_set_nolock(Connection *conn, char *authtype, char *normdn, char
+};
+
+int
+slapi_pblock_set(Slapi_PBlock *pblock, int arg, void *value)
```
I think we still need `PR_ASSERT(NULL != pblock);` here.

mreynolds389 replied:
> I think we still need `PR_ASSERT(NULL != pblock);` here.

Agreed, done!

tbordaz reviewed (approved):
The (huge) change looks good to me and nice.
Minor question, any reason why keeping slapi_pblock_get/set as 'int' rather than int32_t ?

progier389 commented on ldap/servers/slapd/pblock.c:
```diff
@@ -4571,3 +6729,3952 @@ bind_credentials_set_nolock(Connection *conn, char *authtype, char *normdn, char
+{
+    PR_ASSERT(NULL != pblock);
+
+    if (arg > 0 && arg <= PR_ARRAY_SIZE(set_cbtable) && set_cbtable[arg] != NULL) {
```
Should be arg >= 0 && arg<PR_ARRAY_SIZE(set_cbtable) && ... to avoid array overflow ...
(maybe same issue on slapi_pblock_get ?)

mreynolds389 replied:
> Should be arg >= 0 && arg<PR_ARRAY_SIZE(set_cbtable) && ... to avoid array overflow ... (maybe same issue on slapi_pblock_get ?)

My bad, I used the code example _you_ sent me and I didn't fully look it over :-)  I'll get it fixed
